### PR TITLE
Transaction filtering

### LINKED ISSUE #31: <TransactionListContainer /> race condition / unabortable requests.
Currently, on aba7ea6cc3f47f50c5b6dcbbd6db93ec3ef62c27 there is a race condition where the same data gets fetched multiple times. This is how it looks: (note the 2 duplicate groups of 5 transactions)
Peek 2019-04-24 13-13

This is due to these lines of code:
```typescript
// file: /frontend/src/Views/TransactionList/TransactionListContainer.tsx

  const onRequestMoreTranscations = () => {
    dispatch({ type: ActionType.LoadStart, payload: [] });
    // currently fetching relies on transactions being ordered chronologically:
    const lastLoadedTransactionDate =
      state.transactions.length > 0
        ? new Date(state.transactions[state.transactions.length - 1].date)
        : new Date();

    getTransactionListChunk(lastLoadedTransactionDate, 5, filters).then(
      data => {
        if (data.length <= 0) {
          dispatch({ type: ActionType.NoItemsFound, payload: [] });
        } else {
          dispatch({ type: ActionType.ItemsLoaded, payload: data });
        }
      }
    );
  };

  useEffect(() => {
    dispatch({ type: ActionType.ResetTransactionState, payload: [] });
  }, [filters]);
```

What happens is first of all a getTransactionListChunk is called, then the useEffect kicks in, which clears all of the loaded transactions. If the previous request is not completed (which is always the case on my computer), duplicate data gets fetched again.

This can be solved by aborting current requests before dispatching a `ResetTransactionState` action. 
Aborting: https://developers.google.com/web/updates/2017/09/abortable-fetch

### PR DESCRIPTION
**Pull request by Martynas-S**
- Made it possible to filter transactions by Accounts, Categories, Tags and Transaction Types
- When clicking on an account in the account list, it shows all the transactions belonging to the account (automatically applies the account filter)

Peek 2019-04-24 13-45

Everything kind of except for #31, which can be observed in the .gif when I navigate from the accounts to transactions with the account filter. The data is duplicated.

**Need** to fix #31 before merging. Looking for help.
f58f2e3e84b8243f0339d5e6235b18b519209664 is a temporary fix for the race condition on the initial load. Once #31 is fixed f58f2e3e84b8243f0339d5e6235b18b519209664 should be reverted.

**Update**: #31 is fixed

### CODE REVIEW

File: frontend/src/Views/TransactionList/FilterSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  items: FilterItemArray;
+  handleChange: (event: any) => void;
+}
+
+const FilterSelect = ({ label, selected, handleChange, items }: IProps) => {
+  const classes = useStyles();
+
+  const renderValue = (keys: any) =>
+    keys
+      .map((needle: any) => items.find(({ key }) => key === needle)!.value)
```
Comment (Martynas-S) [inline frontend/src/Views/TransactionList/FilterSelect.tsx:47]: https://www.logicbig.com/tutorials/misc/typescript/non-null-assertion-operator.html

**Comment by Martynas-S** [comment]
Right now, each end-consumer of an abortable request **has to** handle abortErrors on their own as such:
```typescript
getTransactionListChunk(lastLoadedTransactionDate, 5, filters, signal)
      .then(data => {
        // handle data
      })
      .catch(e => {
        if (e.name !== "AbortError") {
          throw e;
        }
      });
```

It could be possible to do the generic `AbortError` handling in the http wrapper, but then the return type of making requests would be `Promise<void | Response>` instead of `Promise<Response>` and still changes to the end consumer of request data would have to be made

File: frontend/src/Views/TransactionList/FilterPopupButton.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+import React, { useState } from "react";
+
+interface IProps {
+  children: JSX.Element;
+  className?: string;
+}
+
+const FilterPopupButton = ({ children: Popup, className }: IProps) => {
+  const [anchorEl, setAnchorEl] = useState<any>(null);
```
Comment (Andrius-B) [inline frontend/src/Views/TransactionList/FilterPopupButton.tsx:12]: Neturėtų čia būt `HTMLElement` vietoj `any`?

File: frontend/src/Views/TransactionList/FilterSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  MenuItem,
+  Select
+} from "@material-ui/core";
+import { makeStyles } from "@material-ui/styles";
+import React from "react";
+import { FilterItemArray } from "../../Models/TransactionFilterModel";
+
+const ITEM_HEIGHT = 48;
+const ITEM_PADDING_TOP = 8;
+const MenuProps = {
```
Comment (Andrius-B) [inline frontend/src/Views/TransactionList/FilterSelect.tsx:15]: neišieina šito padayt su `makeStyles`?

File: frontend/src/Views/TransactionList/TransactionFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+const useStyles = makeStyles(() => ({
+  filterPopup: {
+    width: "450px",
+    height: "350px",
+    padding: "20px"
+  }
+}));
+
+type SetStateAction<S> = S | ((prevState: S) => S);
```
Comment (Andrius-B) [inline frontend/src/Views/TransactionList/TransactionFilter.tsx:22]: fancy typescript stuff, nice!

File: frontend/src/Views/TransactionList/TransactionView.tsx [hunk 5]
```diff
@@ -1,41 +1,84 @@
...
+    })),
+    categories: categories.map(({ id, text }) => ({
+      key: id.toString(),
+      value: text
+    })),
+    transactionTypes: ["Expense", "Income", "Transfer"].map(item => ({
+      key: item,
+      value: item
+    })),
+    tags: ["foo", "bar", "baz", "bez", "booze", "bamboozle"].map(item => ({
```
Comment (Andrius-B) [inline frontend/src/Views/TransactionList/TransactionView.tsx:56]: koks čia planas? reiks tag'ų agregatoriaus?

**Review by Andrius-B — APPROVED** [review]

File: frontend/src/Views/TransactionList/TransactionFilter.tsx (hunk 4 above)
Comment (Martynas-S) [reply frontend/src/Views/TransactionList/TransactionFilter.tsx:22]: Nelabai fancy tbh :D

File: frontend/src/Views/TransactionList/TransactionView.tsx (hunk 5 above)
Comment (Martynas-S) [reply frontend/src/Views/TransactionList/TransactionView.tsx:56]: Cia planas toks, kad reikia is kazkur parsitempt tagus. Man atrodo mes neturim dabar tokio dalyko :D

Comment (Andrius-B) [reply frontend/src/Views/TransactionList/TransactionView.tsx:56]: čia šiaip visai heavy-weight operacija gali pasidaryt, dabar tai tikrai nieko panašaus nėra

Comment (Martynas-S) [reply frontend/src/Views/TransactionList/TransactionView.tsx:56]: Reikia pasirasyti issue ;D

File: frontend/src/Views/TransactionList/FilterPopupButton.tsx (hunk 2 above)
Comment (Martynas-S) [reply frontend/src/Views/TransactionList/FilterPopupButton.tsx:12]: Nop, nepaeina

File: frontend/src/Views/TransactionList/FilterSelect.tsx (hunk 3 above)
Comment (Martynas-S) [reply frontend/src/Views/TransactionList/FilterSelect.tsx:15]: Sita parsipuciau is mui docsu: https://material-ui.com/demos/selects/#multiple-select
Kitaip i ta menu neeina perduot props tik per sita, tai del to cia toks kinda clumsy sprendimas

**Pull request merged**